madminer/utils/morphing.py

```diff
@@ -170,23 +252,60 @@ def set_basis(self, basis_from_madminer: Dict[str, Benchmark] = None, basis_nump
             Manually provided morphing matrix. If None, the morphing matrix is calculated automatically. Default value:
             None.
 
+        basis_p : ndarray or None, optional
+            production coupling basis. Default value: None.
+
+        basis_d : ndarray or None, optional
+            decay coupling basis. Default value: None.
+
+        basis_s : ndarray or None, optional
+            couplings that works for both decay and production. Default value: None.
         Returns
         -------
             None
         """
-
-        if basis_from_madminer is not None:
-            self.basis = []
-            for benchmark in basis_from_madminer.values():
-                self.basis.append([benchmark.values[key] for key in self.parameter_names])
-            self.basis = np.array(self.basis)
-        elif basis_numpy is not None:
-            self.basis = np.array(basis_numpy)
+        if basis_p is not None or basis_d is not None or basis_s is not None:
+            # Set the values of basis, basis_p, basis_d, basis_s
+            if basis_s is not None:  
+                self.gs = basis_s
+                self.n_benchmarks = len(basis_s[0])
+
+            if basis_p is not None:
+                self.gp = basis_p
+                self.n_benchmarks = len(basis_p[0])
+            
+            if basis_d is not None:
+                self.gd = basis_d
+                self.n_benchmarks = len(basis_d[0])
+            # Check the which inputs are provided, and set self.n_benchmarks coordinately. 
+            if basis_s is not None and basis_p is not None and basis_d is not None:
+                assert len(basis_p[0]) == len(basis_d[0]) == len(basis_s[0]), "the number of basis points in production, decay and combine should be the same"
```

Comment:
`assert` will be disabled when Python is running with some optimizations, so if you want to use this as a way to inform the user, I would suggest raising an exception (e.g. `raise ValueError("some_text")`).

Reply:
Sure, i will change asserts to exceptions